printing reference interface gives c error: expected struct or union but not 'struct main__IGameObject *'

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 9fc8352, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 5c65e58.9fc8352
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 26100 64-bit
|Processor           |24 cpus, 64bit, little endian, AMD Ryzen 9 3900X 12-Core Processor  
|Memory              |13.38GB/31.9GB
|                    |
|V executable        |C:\v\v.exe
|V last modified time|2025-01-08 08:35:54
|                    |
|V home dir          |OK, value: C:\v
|VMODULES            |OK, value: C:\Users\info\.vmodules
|VTMP                |OK, value: C:\Users\info\AppData\Local\Temp\v_0
|Current working dir |OK, value: D:\vbug2
|                    |
|Git version         |git version 2.42.0.windows.2
|V git status        |weekly.2025.1-12-g9fc83526-dirty
|.git/config present |true
|                    |
|cc version          |N/A
|gcc version         |N/A
|clang version       |clang version 17.0.1
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |thirdparty-windows-amd64 b425ac82
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  main.v && D:\vbug2\main.exe`
```v
module main

interface IGameObject {
mut:
	name     string
	parent   ?&IGameObject
	children []&IGameObject
	add_child(mut o IGameObject)
	advance()
}

@[heap]
struct GameObject implements IGameObject {
mut:
	name     string
	parent   ?&IGameObject
	children []&IGameObject
}

fn (mut gameobject GameObject) add_child(mut o IGameObject) {
	o.parent = &gameobject
	gameobject.children << o
}

fn (mut gameobject GameObject) advance() {
	if gameobject.parent != none {
		eprintln('parent:  ${gameobject.parent.name} ')
	}
	for mut child in gameobject.children {
		child.advance()
	}
}

fn main() {
	mut v1 := &GameObject{
		name: 'v1'
	}
	mut v2 := &GameObject{
		name: 'v2'
	}
	v1.add_child(mut v2)
	v1.advance()
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: C:/Users/info/AppData/Local/Temp/v_0/main.01JH2HA9HEFH6AV8DAFFRW06R8.tmp.c:4050: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/info/AppData/Local/Temp/v_0/main.01JH2HA9HEFH6AV8DAFFRW06R8.tmp.c:6623: error: expected struct or union but not 'struct main__IGameObject *'
... (the original output was 3 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

no c error



> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -158,6 +158,7 @@ mut:
 	inside_for_c_stmt         bool
 	inside_cast_in_heap       int // inside cast to interface type in heap (resolve recursive calls)
 	inside_cast               bool
+	inside_selector           bool
 	inside_memset             bool
 	inside_const              bool
 	inside_array_item         bool
@@ -4008,7 +4009,12 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 					}
 					for i, typ in field.smartcasts {
 						if i == 0 && is_option_unwrap {
-							g.write('(*(${g.styp(typ)}*)')
+							deref := if g.inside_selector {
+								'*'.repeat(field.smartcasts.last().nr_muls() + 1)
+							} else {
+								'*'
+							}
+							g.write('(${deref}(${g.styp(typ)}*)')
 						}
 						g.write('(')
 						if field_sym.kind == .sum_type && !is_option {
@@ -4107,6 +4113,8 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	if field_is_opt {
 		g.write('((${g.base_type(field_typ)})')
 	}
+	old_inside_selector := g.inside_selector
+	g.inside_selector = node.expr is ast.SelectorExpr && node.expr.expr is ast.Ident
 	n_ptr := node.expr_type.nr_muls() - 1
 	if n_ptr > 0 {
 		g.write2('(', '*'.repeat(n_ptr))
@@ -4115,6 +4123,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	} else {
 		g.expr(node.expr)
 	}
+	g.inside_selector = old_inside_selector
 	if field_is_opt {
 		g.write(')')
 	}

--- a/vlib/v/tests/options/option_selector_nested_unwrapped_test.v
+++ b/vlib/v/tests/options/option_selector_nested_unwrapped_test.v
@@ -1,0 +1,44 @@
+module main
+
+interface IGameObject {
+mut:
+	name     string
+	parent   ?&IGameObject
+	children []&IGameObject
+	add_child(mut o IGameObject)
+	advance()
+}
+
+@[heap]
+struct GameObject implements IGameObject {
+mut:
+	name     string
+	parent   ?&IGameObject
+	children []&IGameObject
+}
+
+fn (mut gameobject GameObject) add_child(mut o IGameObject) {
+	o.parent = &gameobject
+	gameobject.children << o
+}
+
+fn (mut gameobject GameObject) advance() {
+	if gameobject.parent != none {
+		eprintln('parent:  ${gameobject.parent.name} ')
+	}
+	for mut child in gameobject.children {
+		child.advance()
+	}
+}
+
+fn test_main() {
+	mut v1 := &GameObject{
+		name: 'v1'
+	}
+	mut v2 := &GameObject{
+		name: 'v2'
+	}
+	v1.add_child(mut v2)
+	v1.advance()
+	assert true
+}

